File: algorithms/game_of_life/game_of_life.js
// Conway's Game of Life - JavaScript implementation

function neighbors(cell) {
  const [x, y] = cell;
  const result = [];
  for (let dx = -1; dx <= 1; dx++) {
    for (let dy = -1; dy <= 1; dy++) {
      if (dx !== 0 || dy !== 0) {
        result.push([x + dx, y + dy]);
      }
    }
  }
  return result;
}


function neighborCount(cell, cells) {
  return neighbors(cell).filter(([x, y]) => cells.has(`${x},${y}`)).length;
}


function step(cells) {
  const candidates = new Set();
  
  // Add all live cells and their neighbors to candidates
  for (const key of cells) {
    candidates.add(key);
    const [x, y] = key.split(',').map(Number);
    neighbors([x, y]).forEach(([nx, ny]) => candidates.add(`${nx},${ny}`));
  }
  
  const nextGen = new Set();
  for (const key of candidates) {
    const [x, y] = key.split(',').map(Number);
    const count = neighborCount([x, y], cells);
    if ((cells.has(key) && (count === 2 || count === 3)) || 
        (!cells.has(key) && count === 3)) {
      nextGen.add(key);
    }
  }
  
  return nextGen;
}


function life(cells, generations) {
  for (let i = 0; i < generations; i++) {
    cells = step(cells);
  }
  return cells;
}


// Export for Node.js
if (typeof module !== 'undefined' && module.exports) {
  module.exports = { neighbors, neighborCount, step, life };
}

File: algorithms/game_of_life/game_of_life.rb
# Conway's Game of Life - Ruby implementation
require 'set'

def neighbors(cell)
  x, y = cell
  (-1..1).flat_map { |dx| (-1..1).map { |dy| [x+dx, y+dy] if dx != 0 || dy != 0 } }.compact.to_set
end


def neighbor_count(cell, cells)
  neighbors(cell).count { |n| cells.include?(n) }
end


def step(cells)
  candidates = cells | cells.flat_map { |cell| neighbors(cell) }
  candidates.select do |cell|
    count = neighbor_count(cell, cells)
    (cells.include?(cell) && (count == 2 || count == 3)) || (!cells.include?(cell) && count == 3)
  end.to_set
end


def life(cells, generations)
  generations.times { cells = step(cells) }
  cells
end
